nemo/collections/nlp/modules/common/transformer/transformer_encoders.py

```diff
@@ -0,0 +1,105 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+
+import torch
+import torch.nn as nn
+
+from nemo.collections.common.parts import form_attention_mask
+from nemo.collections.nlp.modules.common.transformer.transformer_modules import MultiHeadAttention, PositionWiseFF
+
+__all__ = ["TransformerEncoder"]
+
+
+class TransformerEncoderBlock(nn.Module):
+    """
+    Building block of Transformer encoder.
+
+    Args:
+        hidden_size: size of the embeddings in the model, also known as d_model
+        inner_size: number of neurons in the intermediate part of feed-forward
+            net, usually is (4-8 x hidden_size) in the papers
+        num_attention_heads: number of heads in multi-head attention
+        attn_score_dropout: probability of dropout applied to attention scores
+        attn_layer_dropout: probability of dropout applied to the output of the
+            attention layers, but before layer normalization
+        ffn_dropout: probability of dropout applied to FFN output
+        hidden_act: activation function used between two linear layers in FFN
+    """
+
+    def __init__(
+        self,
+        hidden_size,
+        inner_size,
+        num_attention_heads=1,
+        attn_score_dropout=0,
+        attn_layer_dropout=0,
+        ffn_dropout=0,
+        hidden_act="relu",
+    ):
+        super().__init__()
+
+        self.first_sub_layer = MultiHeadAttention(
+            hidden_size, num_attention_heads, attn_score_dropout, attn_layer_dropout
+        )
+        self.second_sub_layer = PositionWiseFF(hidden_size, inner_size, ffn_dropout, hidden_act)
+
+    def forward(self, encoder_query, encoder_mask, encoder_keys):
+        self_attn_output = self.first_sub_layer(encoder_query, encoder_keys, encoder_keys, encoder_mask)
+        output_states = self.second_sub_layer(self_attn_output)
+        return output_states
+
+
+class TransformerEncoder(nn.Module):
```

Comment:
Could you please inherit this from NeuralModule instead of nn.Module, define input/output types and add @typecheck() to forward. 
You should not have to change anything else, might need to add some empty implementations for save/restore for now.

Reply:
I tried to add input/output types to TransformerEncoder but encountered the following problem. My forward returns either list of Tensors or a single Tensor depending on the boolean argument return_mems. The former regime is important for sequence generation (e.g. via beam search) and the latter is used during training. If I set output type to Void, it fails to typecheck with the input to TokenClassifier which I use after encoder; if I set output type to NeuralType(('B', 'T', 'D'), ChannelType()) it works perfectly fine during training but I can not use it when return_mems=True.

Is there a way to save required flexibility without completely rewriting my classes or splitting them into separate classes?

Also, I am wondering whether it is really necessary to force typechecking in all our models. I'd rather prefer writing native Pytorch classes and wrapping them into ModelPT as I did in transformer_lm_model.py. For example, it allows to easily port such code to other projects which do not use NeMo.

Reply:
I can add input/output types to transformer_lm_model.py though